patches/tModLoader/Terraria/ModLoader/Core/SourceManagement.cs

```diff
@@ -0,0 +1,320 @@
+		if (!File.Exists(Path.Combine(modSrcDirectory, "Properties", "launchSettings.json"))) {
+			modifications.Add(() => TryWriteModTemplateFile(modSrcDirectory, "Properties/launchSettings.json", boxedParameters));
+		}
+
+		// Do some cleanups, but only if we already have something.
+		if (modifications.Count != 0) {
+			try {
+				// Old files can cause some issues.
+				DeleteIfExists(new DirectoryInfo(Path.Combine(modSrcDirectory, "obj")));
+				DeleteIfExists(new DirectoryInfo(Path.Combine(modSrcDirectory, "bin")));
+
+				//TODO: Why do we do this?
+				DeleteIfExists(new FileInfo(Path.Combine(modSrcDirectory, "Properties", "AssemblyInfo.cs")));
+			}
+			catch { }
+		}
+
+		return modifications;
+	}
+
+	/// <summary> Checks a mod source-code directory for available upgrades, optionally applying them. </summary>
+	private static bool TryGetCsprojUpgradeAction(string modSrcDirectory, [NotNullWhen(true)] out Action? result, object? templateParameters = null)
+	{
+		var status = UpgradeStatus.NotRequired;
+		string csprojPath = Path.Combine(modSrcDirectory, $"{Path.GetFileName(modSrcDirectory)}.csproj");
+		
+		// Check if the file even exists.
+		if (!File.Exists(csprojPath))
+			status = UpgradeStatus.FileMissing;
+
+		// Try to load the XML document.
+		XDocument? document = null;
+		if (status is not UpgradeStatus.FileMissing && !TryLoadXmlDocument(csprojPath, LoadOptions.PreserveWhitespace, out document)) {
+			status = UpgradeStatus.FileBroken;
+		}
+
+		// Try to collect required modifications.
+		var modifications = new List<Action>();
+		if (document != null && !TryCollectingCsprojModifications(document, modifications)) {
+			status = UpgradeStatus.FileBroken;
+		}
+
+		// If there are any modifications - the file is upgradeable.
+		if (modifications?.Count > 0) {
+			status = UpgradeStatus.Upgradeable;
+		}
+
+		// If this is not a dry run and there's changes to be made.
+		if (status is not UpgradeStatus.NotRequired) {
+			result = () => {
+				// Apply modifications if any exist.
+				modifications!.ForEach(m => m());
+
+				// Recreate the file from scratch if it's missing or broken, otherwise just save our XML.
+				if (status is UpgradeStatus.FileMissing or UpgradeStatus.FileBroken) {
+					ResetCsprojFile(csprojPath, createBackup: status is not UpgradeStatus.FileMissing, templateParameters);
+				} else {
+					WriteXmlDocumentToFile(csprojPath, document!);
+				}
+			};
+
+			return true;
+		}
+
+		result = default;
+		return false;
+	}
+
+	/// <summary> If succesfful - populates the provided list with delegates, executing which will upgrade the provided document. </summary>
+	private static bool TryCollectingCsprojModifications(XDocument document, List<Action> modifications)
+	{
+		// Check if this is even a C# project.
+		if (document!.Root is not { Name.LocalName: "Project", FirstAttribute: { Name.LocalName: "Sdk", Value: "Microsoft.NET.Sdk" } } root) {
+			return false;
```

Comment:
this is currently the only way to get `FileBroken`. See my above comment about factoring this out, so this method doesn't need a `bool` return, and can just return (or `yield`) the list of modifications instead